### ISSUE TITLE
Lacking guide to production environment deployment

### ISSUE THREAD

**Issue by fire** [access-company/antikythera#73]
https://hexdocs.pm/antikythera/development_environment.html

Development guides for antikythera exist but there lacks a guide for production deployments.

What cicd is needed to deploy?

How does the antikythera code get merged with the gears code in a docker image?

How does ACCESS deploy: VMs? Docker? Kubernetes?

What environmental variables and other settings are required?

**Comment by aMasakiTakahashi** [comment]
Sorry, we don't have guides for production deployments.
We will add the necessary documents in the future.

For your reference, we(ACCESS company) briefly introduce how we construct Antikythera cluster for production.
We have built Antikythera on AWS. (we don't use docker and so on, so currently build each Erlang node on EC2(Amazon Elastic Compute Cloud) directly).

Each Erlang VM nodes runs on EC2, Furthermore, we use an EC2 node for Jenkins server and use EFS(Amazon Elastic File System).
Jenkins server checks gear code repositories periodically.
If it finds a new version of gear, it compiles gear and stores the compiled object on EFS.
Each Erlang VM node monitors EFS periodically. (This is implemented by [VersionSynchronizer](https://github.com/access-company/antikythera/blob/master/core/sup_tree_core/version_synchronizer.ex)
If a node finds the new compiled object of gear, It [loads](https://github.com/access-company/antikythera/blob/master/core/version/gear.ex#L52) new version of gear with `hot code upgrading`.

Upgrading mechanism of Antikythera core is also the same as gears.

[Note]
Antikythera has been designed not to depend on the specific cloud provider.
In other words, modules depending on some infrastructure are completely divided from Antikythera core codes.
We offer these modules as pluggable modules of Antikythera.
AWS Example is [here](https://github.com/access-company/antikythera_aws).

>What environmental variables and other settings are required?

You have to setup build server(ex, Jenkins) to compile gear and file server(ex, EFS) to store compile gear object.

**Comment by fire** [comment]
You can close this when documentation is written.

**Comment by aMasakiTakahashi** [comment]
>Are you using EFS or S3?

We use EFS, but Antikythera doesn't depend on EFS.
Antikythera requires only a file server.
Each ErlangVM node should be mounted on a file server.
But It is OK that this file server is not EFS.

>A common problem converting vms to docker / kubernetes is that secrets need to be moved into environment variables, so everything configuration wise that is written to the executable needs to be changed.
>For example a previous example would have javascript that encodes the endpoint of the elixir server. When the server endpoint is changed, it causes the javascript to fail.

We use Jenkins to build Antikythera and embed some secret info at the same time.
After that, we move compiled Antikythera to linux server and run it.
So, I think you can control settings of Antikythera with environmental variables.

**Comment by aMasakiTakahashi** [comment]
>In Kubernetes I can give each ErlangVM a disk but they are separate from each other and small. There can be a way to reclaim disk space, but it is not access to a file system that has near unlimited size.

Currently, Antikythera need file system to share some files with each node.
I think that it is also good to use S3 instead of a file server.
But our priority to support S3 for that purpose is low.

>But It is OK that this file server is not EFS.
>What do you mean?

It means that Antikythera doesn't depend on AWS service.
